src/dev/encoder.cpp

```diff
@@ -0,0 +1,141 @@
+    recentBInterruptTime = 0;
+}
+
+void Encoder::aInterruptHandler(IO::GPIO* pin) {
+    if ((time::millis() - recentAInterruptTime) < INTERRUPTCOOLDOWN) {
+        log::LOGGER.log(log::Logger::LogLevel::DEBUG, "aInterrupt Skipped, reason: on cooldown");
+        return;
+    }
+    int8_t change;
+    switch (currentRelPos) {
+    case 0:
+        change = 1;
+        currentRelPos = 1;
+        break;
+    case 1:
+        change = -1;
+        currentRelPos = 0;
+        break;
+    case 2:
+        change = 1;
+        currentRelPos = 3;
+        break;
+    case 3:
+        change = -1;
+        currentRelPos = 2;
+        break;
+    }
+    interruptChange += change;
+    recentAInterruptTime = time::millis();
+    //time::wait(200);
+    log::LOGGER.log(log::Logger::LogLevel::DEBUG, "aInterrupt Called, "
+                                                  "\n\r\tCalculated Position: %d"
+                                                  "\n\r\tActual Position: %d"
+                                                  "\n\r\tChange: %d",
+                    currentRelPos, readPinValues(), change);
+}
+
+void Encoder::bInterruptHandler(IO::GPIO* pin) {
+    if ((time::millis() - recentBInterruptTime) < INTERRUPTCOOLDOWN) {
+        log::LOGGER.log(log::Logger::LogLevel::DEBUG, "bInterrupt Skipped, reason: on cooldown");
+        return;
+    }
+    int8_t change;
+    switch (currentRelPos) {
+    case 0:
+        change = -1;
+        currentRelPos = 3;
+        break;
+    case 1:
+        change = 1;
+        currentRelPos = 2;
+        break;
+    case 2:
+        change = -1;
+        currentRelPos = 1;
+        break;
+    case 3:
+        change = 1;
+        currentRelPos = 0;
+        break;
+    }
+    interruptChange += change;
+    recentBInterruptTime = time::millis();
+    log::LOGGER.log(log::Logger::LogLevel::DEBUG, "bInterrupt Called, "
+                                                  "\n\r\tCalculated Position: %d"
+                                                  "\n\r\tActual Position: %d"
+                                                  "\n\r\tChange: %d",
+                    currentRelPos, readPinValues(), change);
+}
+
+uint64_t Encoder::getPosition() const {
+    return position;
+}
+
+void Encoder::update() {
+    int64_t change = interruptChange;
+    interruptChange = 0;
+    changePosition(change);
+    //in case of triggers desyncing the position
+    currentRelPos = readPinValues();
+}
+
+int8_t Encoder::readPinValues() {
+    bool aPos = (bool) a.readPin();
+    bool bPos = (bool) b.readPin();
+    int8_t newPos = convertPinValuesToPosition(aPos, bPos);
+    return newPos;
+}
+
+int8_t Encoder::convertPinValuesToPosition(bool a, bool b) {
+    if (a == 0 && b == 0) {
+        return 0;
+    } else if (a == 1 && b == 0) {
+        return 1;
+    } else if (a == 1 && b == 1) {
+        return 2;
+    } else if (a == 0 && b == 1) {
+        return 3;
+    }
+}
+
+bool Encoder::changePosition(int64_t change) {
```

Comment:
This function will break if you are using ranges, positions, etc that are greater than 2^63. While that is unlikely, you should rework this, so that it doesn't have that issue

Reply:
fixed!